src/virt.rs

```diff
@@ -508,15 +506,27 @@ impl VirtContext {
                         instr
                     );
                     self.handle_device_access_fault(&instr, &device);
+                } else if (self.csr.mstatus & mstatus::MPRV_FILTER) >> mstatus::MPRV_OFFSET == 1 {
+                    // Can an access to device be legally requested with a virtual address?
```

Comment:
I think so yes. Is the address in `trap_info.mtval` a virtual address in this case?

Reply:
Yes, one example would be `ff60000001d14076`.
Then we should somehow check if the virtual address corresponds to any device, and use device handler if that's the case (I don't have any ideas on how to do that though)